core/src/main/java/org/apache/iceberg/SnapshotSummary.java

```diff
@@ -161,27 +176,154 @@ public void merge(SnapshotSummary.Builder builder) {
-      setIf(removedEqDeletes > 0, builder, REMOVED_EQ_DELETES_PROP, removedEqDeletes);
-      setIf(true, builder, CHANGED_PARTITION_COUNT_PROP, changedPartitions.size());
 
-      return builder.build();
+      if (trustSizeAndDeleteCounts) {
+        setIf(addedSize > 0, builder, ADDED_FILE_SIZE_PROP, addedSize);
+        setIf(removedSize > 0, builder, REMOVED_FILE_SIZE_PROP, removedSize);
+        setIf(addedPosDeletes > 0, builder, ADDED_POS_DELETES_PROP, addedPosDeletes);
+        setIf(removedPosDeletes > 0, builder, REMOVED_POS_DELETES_PROP, removedPosDeletes);
+        setIf(addedEqDeletes > 0, builder, ADDED_EQ_DELETES_PROP, addedEqDeletes);
+        setIf(removedEqDeletes > 0, builder, REMOVED_EQ_DELETES_PROP, removedEqDeletes);
+      }
     }
 
-    private static void setIf(boolean expression, ImmutableMap.Builder<String, String> builder,
-                              String property, Object value) {
-      if (expression) {
-        builder.put(property, String.valueOf(value));
+    void addedFile(ContentFile<?> file) {
+      this.addedSize += file.fileSizeInBytes();
+      switch (file.content()) {
+        case DATA:
+          this.addedFiles += 1;
+          this.addedRecords += file.recordCount();
+          break;
+        case POSITION_DELETES:
+          this.addedDeleteFiles += 1;
+          this.addedPosDeletes += file.recordCount();
+          break;
+        case EQUALITY_DELETES:
+          this.addedDeleteFiles += 1;
+          this.addedEqDeletes += file.recordCount();
+          break;
+        default:
+          throw new UnsupportedOperationException("Unsupported file content type: " + file.content());
       }
     }
+
+    void removedFile(ContentFile<?> file) {
+      this.removedSize += file.fileSizeInBytes();
+      switch (file.content()) {
+        case DATA:
+          this.removedFiles += 1;
+          this.deletedRecords += file.recordCount();
+          break;
+        case POSITION_DELETES:
+          this.removedDeleteFiles += 1;
+          this.removedPosDeletes += file.recordCount();
+          break;
+        case EQUALITY_DELETES:
+          this.removedDeleteFiles += 1;
+          this.removedEqDeletes += file.recordCount();
+          break;
+        default:
+          throw new UnsupportedOperationException("Unsupported file content type: " + file.content());
+      }
+    }
+
+    void addedManifest(ManifestFile manifest) {
+      switch (manifest.content()) {
+        case DATA:
+          this.addedFiles += manifest.addedFilesCount();
+          this.addedRecords += manifest.addedRowsCount();
+          this.removedFiles += manifest.deletedFilesCount();
+          this.deletedRecords += manifest.deletedRowsCount();
+          break;
+        case DELETES:
+          this.addedDeleteFiles += manifest.addedFilesCount();
+          this.removedDeleteFiles += manifest.deletedFilesCount();
+          this.trustSizeAndDeleteCounts = false;
```

Comment:
Is this because we don't know how many records an equality delete removes?

Reply:
Actually, no. It is because we don't have row counts for the number of equality and positional deletes in delete manifests.